### PR TITLE
[receiver/awss3receiver]: Add ingest progress notifications via OpAMP

### PR DESCRIPTION
**Pull request by adcharre**
**Description:** Add support for monitoring the progress of ingesting data from an S3 bucket via OpAMP custom messages.
This change would allow an appropriate UI to display the state of ingested.

**Link to tracking Issue:** 30750

**Testing:** Additional unit tests added. Full stack testing using a built collector and the example OpAMP server provided in the PR.

**Documentation:** README.md updated to include details of the configuration changes and details of the OpAMP Custom Capability and message format.

### CODE REVIEW

File: .chloggen/awss3receiver_notifications.yaml [hunk 1]
```diff
@@ -0,0 +1,27 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: 'new_component'
```
Comment (atoulme) [inline .chloggen/awss3receiver_notifications.yaml:4]:
```suggestion
change_type: 'enhancement'
```

File: .chloggen/awss3receiver_notifications.yaml [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: awss3receiver
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: 'Add support for monitoring the progress of ingesting data from an S3 bucket via OpAMP custom messages.'
```
Comment (atoulme) [inline .chloggen/awss3receiver_notifications.yaml:10]: wow. That's the first time I see this done. Can you use obsreport to report metrics first?

File: receiver/awss3receiver/README.md [hunk 3]
```diff
@@ -46,4 +48,23 @@ receivers:
...
+The format of the notifications is a JSON object with the following fields:
+
+| Field             | Description                                                                     |
+|:------------------|:--------------------------------------------------------------------------------|
+| `telemetry_type`  | The type of telemetry being ingested. One of "traces", "metrics", or "logs".    |
+| `ingest_status`   | The status of the data ingestion. One of "ingesting", "failed", or "completed". |
+| `start_time`      | The time to start retrieving data in RFC3339 format.                            |
+| `stop_time`       | The time to stop retrieving data in RFC3339 format.                             |
+| `ingest_time`     | The time of the data currently being ingested in RFC3339 format.                |
+| `failure_message` | Error message if `ingest_status` is "failed".                                   |
```
Comment (atoulme) [inline receiver/awss3receiver/README.md:65]: Is there a spec for this as part of opAmp or maybe entities?

File: receiver/awss3receiver/notifications_test.go [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+	"github.com/open-telemetry/opamp-go/client/types"
+	"github.com/open-telemetry/opamp-go/protobufs"
+	"github.com/stretchr/testify/require"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/component/componenttest"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/extension/opampcustommessages"
+)
+
+type mockCustomCapabilityRegistry struct {
```
Comment (atoulme) [inline receiver/awss3receiver/notifications_test.go:22]: shouldn't this be in the test file?

File: receiver/awss3receiver/notifications_test.go [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+	unregisterCalled bool
+	sentMessages     []customMessage
+}
+
+type customMessage struct {
+	messageType string
+	message     []byte
+}
+
+type hostWithExtensions struct {
```
Comment (atoulme) [inline receiver/awss3receiver/notifications_test.go:38]: this should be in the test file, I think

File: receiver/awss3receiver/notifications.go [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+	handler, err := registry.Register(CustomCapability)
+	if err != nil {
+		return fmt.Errorf("failed to register custom capability: %w", err)
+	}
+	n.handler = handler
+	return nil
+}
+
+func (n *opampNotifier) Shutdown(_ context.Context) error {
```
Comment (atoulme) [inline receiver/awss3receiver/notifications.go:72]: Why are you exporting Shutdown, Start and SendStatus? Are all 3 part of an interface?

Comment (adcharre) [reply receiver/awss3receiver/notifications.go:72]: They are part of the `statusNotifier` interface defined in the same file.  Possibly overkill at this stage, I was allowing for other methods of notification (ie Webhook, file ...)

File: receiver/awss3receiver/README.md (hunk 3 above)
Comment (adcharre) [reply receiver/awss3receiver/README.md:65]: There is no spec for this and that is the reason why I'm using a Custom Capability.

**Comment by atoulme** [comment]
I believe it is worth a good SIG convo:
When is that advisable?
Looks like you'd like metrics for this, right?
Or maybe even traces?
Is opAmp the place to send notifications too anyway? Is there a spec, is there something else here I'm missing? Can we do this on a more established component first to test it out with more eyeballs?

**Comment by adcharre** [comment]
@atoulme  / @tigrannajaryan  - Fair questions - though I'm not looking for metrics as the intention is to use the collector as one component of a larger app to retrieve data from S3 back into an observability platform. As a result I'm looking at being able to report progress to a user so they know when the requested trace has been ingested. 

I'll look at adding obsreport metrics as suggested while this is up for discussion.

**Comment by tigrannajaryan** [comment]
@adcharre it would be great to learn more about your use case and perhaps OpAMP is the right solution for it.

**Comment by atoulme** [comment]
We're discussing this here as well: https://cloud-native.slack.com/archives/C02J58HR58R/p1715714974462069

File: receiver/awss3receiver/README.md [hunk 7]
```diff
@@ -46,4 +48,23 @@ receivers:
...
+
+The "ingesting" status is sent at the beginning of the ingest process before data has been retrieved for the specified time.
+If during the processing of the data an error occurs a status message with `ingest_status` set to "failed" status with 
+the time of the data being ingested when the failure occurred.
+If the ingest process completes successfully a status message with `ingest_status` set to "completed" is sent.
```
Comment (tiffany76) [inline receiver/awss3receiver/README.md:70]: Docs writer here. Just a quick edit for clarity. Thanks!

```suggestion
If an error occurs during the processing of the data, a status message with `ingest_status` set to "failed" is sent, along with the time the ingestion failure occurred.
If the ingest process completes successfully, a status message with `ingest_status` set to "completed" is sent.
```

**Pull request closed without merging**